src/Rule/CompareHandler.php

```diff
@@ -77,23 +80,68 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
      */
     private function compareValues(string $operator, string $type, mixed $value, mixed $targetValue): bool
     {
-        if ($type === CompareType::NUMBER) {
-            $value = (float) $value;
-            $targetValue = (float) $targetValue;
-        } else {
-            $value = (string) $value;
-            $targetValue = (string) $targetValue;
+        if (!in_array($operator, ['==', '===', '!=', '!=='])) {
+            [$value, $targetValue] = $this->getTypeCastedValues($type, $value, $targetValue);
         }
 
         return match ($operator) {
-            '==' => $value == $targetValue,
-            '===' => $value === $targetValue,
-            '!=' => $value != $targetValue,
-            '!==' => $value !== $targetValue,
+            '==' => $this->assertEquals($type, $value, $targetValue),
+            '===' => $this->assertEquals($type, $value, $targetValue, strict: true),
+            '!=' => $this->assertNotEquals($type, $value, $targetValue),
+            '!==' => $this->assertNotEquals($type, $value, $targetValue, strict: true),
             '>' => $value > $targetValue,
             '>=' => $value >= $targetValue,
             '<' => $value < $targetValue,
             '<=' => $value <= $targetValue,
         };
     }
+
+    private function assertEquals(
```

Comment:
```suggestion
    private function equals(
```

because it's not a usual assertion throwing an exception on negative result of a check.

Reply:
I renamed it, but slightly different, because suggested version doesn't imply a question.